librz/analysis/p/analysis_avr.c

```diff
@@ -1662,16 +1662,16 @@ static int avr_op(RzAnalysis *analysis, RzAnalysisOp *op, ut64 addr, const ut8 *
 		offset = 0;
 		rz_analysis_esil_reg_write(analysis->esil, "_prog", offset);
 
-		offset += (1 << cpu->pc);
+		offset += (1 << analysis->arch_target->profile->pc);
 		rz_analysis_esil_reg_write(analysis->esil, "_io", offset);
 
-		offset += const_get_value(const_by_name(cpu, CPU_CONST_PARAM, "sram_start"));
+		offset += analysis->arch_target->profile->sram_start;
 		rz_analysis_esil_reg_write(analysis->esil, "_sram", offset);
 
-		offset += const_get_value(const_by_name(cpu, CPU_CONST_PARAM, "sram_size"));
+		offset += analysis->arch_target->profile->sram_size;
 		rz_analysis_esil_reg_write(analysis->esil, "_eeprom", offset);
 
-		offset += const_get_value(const_by_name(cpu, CPU_CONST_PARAM, "eeprom_size"));
+		offset += analysis->arch_target->profile->eeprom_size;
 		rz_analysis_esil_reg_write(analysis->esil, "_page", offset);
```

Comment:
what does happen here if a unsupported CPU is given?

Reply:
Great catch. I am not sure, actually.
The situation was same before, as well: wrong values (zeroes) gets written.
What do you think we can do about it?

Reply:
We could create "generic" profile, maybe?

Reply:
Yeah, that sounds like a plan!
 
Such a check can be done near where the `path` is initialized in `rz_arch_profiles_init`.
```c
if (CPU is unsupported) {
      switch (asmarch) {
      case AVR:
            path = path_to_avr_generic_profile;
}
```
Or something similar can be done.

So, how do I check whether a CPU is unsupported?

Reply:
I suppose when you load SDBs you could store all CPU names that are loaded and store it in some vector.

Reply:
i think we should not support this for unknown cpus.
we can just throw an error message saying that there is no profile available for this cpu. and just initialize an empty structure and avoid enriching the data

Reply:
I think we might need to move the SDB loading of Arch Profiles to RzCore/RzAnalysis initialization before doing this.

Because, while I was doing this, I saw that `x86` was passed as the `arch` and `cpu at the beginning and I think profile's `init` (basically, `cb_asmarch` and `cb_asmcpu`) is called so many times before the actual `arch` and `cpu` are passed/known.

Reply:
sounds an interesting issue. what you could do is make a branch and see what happens if you load this under rzanalysis; then if it solves the issue you can join the diverged branch here and push it.

Reply:
Well actually, I've managed to hack it.
Since the diff is not that big, I'm just gonna post the diff here and see if that's worthy enough to be added.

Printing an error message makes it print a couple of times, so I'm just loading ATMega8's (probably the most popular) profile here if the CPU is not valid.

```diff
diff --git a/librz/analysis/arch_profile.c b/librz/analysis/arch_profile.c
index 575290b7d..ec35782ef 100644
--- a/librz/analysis/arch_profile.c
+++ b/librz/analysis/arch_profile.c
@@ -149,6 +149,43 @@ RZ_API bool rz_type_db_load_arch_profile_sdb(RzArchTarget *t, const char *path)
 	return sdb_load_arch_profile_by_path(t, path);
 }
 
+static bool is_cpu_valid(char *cpu_dir, const char *cpu) {
+	RzList *files = rz_sys_dir(cpu_dir);
+	if (!files) {
+		return false;
+	}
+	RzListIter *it;
+	char *filename = NULL;
+	char *cpu_name = NULL;
+	char *arch_cpu = NULL;
+	RzPVector *cpus = rz_pvector_new(NULL);
+	rz_pvector_init(cpus, NULL);
+
+	rz_list_foreach (files, it, filename) {
+		if (!strcmp(filename, "..") || !strcmp(filename, "..")) {
+			continue;
+		}
+		arch_cpu = rz_str_ndup(filename, strlen(filename) - 4);
+		if (!arch_cpu)
+			continue;
+		cpu_name = strchr(arch_cpu, '-');
+		rz_str_remove_char(cpu_name, '-');
+		rz_pvector_push(cpus, cpu_name);
+	}
+	rz_list_free(files);
+	free(arch_cpu);
+
+	void **ita;
+	rz_pvector_foreach (cpus, ita) {
+		if (!strcmp(*ita, cpu)) {
+			rz_pvector_free(cpus);
+			return true;
+		}
+	}
+	rz_pvector_free(cpus);
+	return false;
+}
+
 /**
  * \brief Initializes RzArchProfile by loading the path to the SDB file
  * 		  of the CPU profile
@@ -167,10 +204,19 @@ RZ_API bool rz_arch_profiles_init(RzArchTarget *t, const char *cpu, const char *
 	if (!path) {
 		return false;
 	}
+	char *cpu_dir = rz_str_newf(RZ_JOIN_3_PATHS("%s", RZ_SDB, "asm/cpus"), dir_prefix);
+	if (!is_cpu_valid(cpu_dir, cpu)) {
+		if (arch) {
+			if (!strcmp(arch, "avr")) {
+				path = rz_str_newf(RZ_JOIN_4_PATHS("%s", RZ_SDB, "asm/cpus", "avr-ATmega8.sdb"), dir_prefix);
+			}
+		}
+	}
 	if (!rz_type_db_load_arch_profile_sdb(t, path)) {
 		sdb_free(t->db);
 		t->db = NULL;
 	}
 	free(path);
+	free(cpu_dir);
 	return true;
 }
\ No newline at end of file
```

Reply:
looks like a good patch, just check that `RzPVector *cpus` is not NULL